File: javascript/programmers/javascript/level2/조이스틱.js
function solution(name) {
  const start = `A`.charCodeAt();
  const end = `Z`.charCodeAt() + 1;
  const strs = name.split(``);
  let answer = 0;
  let i = 0;

  while (true) {
    if (strs[i] !== `A`) {
      const cur = strs[i].charCodeAt();
      const down = Math.abs(start - cur);
      const up = Math.abs(end - cur);

      answer += Math.min(down, up);
      strs[i] = `A`;
    }
    // 알파벳 이름을 완성하였을 때는 커서를 이동할 필요가 없음
    if (strs.findIndex((str) => str !== `A`) === -1) {
      break;
    }
    let rightCount = 0;
    let rightIndex = 0;
    let leftCount = 0;
    let leftIndex = 0;

    // 오른 쪽 카운트
    for (let right = 1; right < strs.length; right++) {
      rightIndex = (i + right) % strs.length;
      if (strs[rightIndex] !== `A`) {
        rightCount = right;
        break;
      }
    }
    // 왼 쪽 카운트
    for (let left = 1; left < strs.length; left++) {
      leftIndex = i - left < 0 ? i - left + strs.length : i - left;
      if (strs[leftIndex] !== `A`) {
        leftCount = left;
        break;
      }
    }
    // 더 가까운 쪽으로 이동
    if (rightCount <= leftCount) {
      i = rightIndex;
      answer += rightCount;
    } else {
      i = leftIndex;
      answer += leftCount;
    }
  }
  return answer;
}

File: python/programmers/level2/조이스틱.py
def solution(name):
    arr = ['A'] * len(name)
    i = 0
    ans = 0
    while True:
        # 현재 위치의 알파벳을 name 알파벳과 동일하게 바꾼다.
        ans += min(ord(name[i]) - ord("A"), ord("Z") - ord(name[i]) + 1)
        arr[i] = name[i]
        # 만약 알파벳 하나를 바꾼 후 name 과 일치한다면 종료
        if ''.join(map(str, arr)) == name:
            break
        # 바꾸는 곳의 가장 가까운 다음 알파벳으로 간다.
        left, right = 1, 1
        while name[i - left] == arr[i - left]:
            left += 1
        while name[i + right] == arr[i + right]:
            right += 1
        if left < right:
            i = i - left
            ans += left
        else:
            i = i + right
            ans += right
        # print(''.join(map(str, arr)))
        # print(ans)
    return ans


print(solution("JEROEN"))
